### PR TITLE
[SPARK-41108][CONNECT] Control the max size of arrow batch

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

Control the max size of arrow batch

### Why are the changes needed?

as per the suggestion https://github.com/apache/spark/pull/38468#discussion_r1018951362

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala [hunk 1]
```diff
@@ -161,17 +166,23 @@ private[sql] object ArrowConverters extends Logging {
...
         Utils.tryWithSafeFinally {
-          while (rowIter.hasNext && (maxRecordsPerBatch <= 0 || rowCount < maxRecordsPerBatch)) {
+          // Always write the schema.
+          MessageSerializer.serialize(writeChannel, arrowSchema)
+
+          // Always write the first row.
+          while (rowIter.hasNext && (rowCount == 0 || estimatedBatchSize < maxBatchSize)) {
             val row = rowIter.next()
             arrowWriter.write(row)
+            estimatedBatchSize += row.asInstanceOf[UnsafeRow].getSizeInBytes
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:178]: refer to how the size is computed in [BroadcastExchange](https://github.com/apache/spark/blob/e17d8ecabcad6e84428752b977120ff355a4007a/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L150-L158) 

but not 100% sure, should I use this instead?
```
row match {
   case unsafe: UnsafeRow => estimatedBatchSize += unsafe.getSizeInBytes
   case _ => estimatedBatchSize += SizeEstimator.estimate(row)
}
```

cc @HyukjinKwon

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:178]: The size of message should be based on Arrow but we are only able to know the size of the batch when Arrow batch is created.

So I am fine with the current approach. I do believe that `UnsafeRow` has bigger size than `ArrowBatch` in general.

One nit would be we should probably set the lower size in `maxBatchSize` to be conservative. For example, `maxBatchSize * 0.7`

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala (hunk 1 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:178]: will update `maxBatchSize * 0.7`

**Review by amaliujia — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Let me actually merge and refactor this out. I am working on it actually.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**